clang/lib/APINotes/APINotesWriter.cpp

```diff
@@ -56,15 +56,21 @@ class APINotesWriter::Implementation {
 
   bool SwiftInferImportAsMember = false;
 
-  /// Information about Objective-C contexts (classes or protocols).
+  /// Information about contexts (Objective-C classes or protocols or C++
+  /// namespaces).
   ///
-  /// Indexed by the identifier ID and a bit indication whether we're looking
-  /// for a class (0) or protocol (1) and provides both the context ID and
-  /// information describing the context within that module.
-  llvm::DenseMap<std::pair<unsigned, char>,
+  /// Indexed by the parent context ID, context kind and the identifier ID of
+  /// this context and provides both the context ID and information describing
+  /// the context within that module.
+  llvm::DenseMap<std::tuple<uint32_t, uint8_t, uint32_t>,
```

Comment:
One suggestion might be to make this a struct so you can access the fields by their name instead of relying on documentation to remember which field is which :)

Reply:
Every time I use `std::tuple` in a data structure, I regret it eventually ;)

Reply:
Agreed, replaced the tuples with a struct `ContextTableKey`.